crates/subspace-transaction-pool/src/bundle_validator.rs

```diff
@@ -0,0 +1,256 @@
+        let pre_best_hash = self.best_hash();
+        if let Err(err) = self.on_import_new_best_block(new_best_hash) {
+            tracing::error!(
+                %err,
+                "Failed to handle new best block for bundle-collector"
+            );
+        }
+        if pre_best_hash != self.best_hash() {
+            self.bundle_syncer
+                .update(self.bundle_stored_in_last_k.clone());
+        }
+    }
+}
+
+/// `BundleStoredInLastK` used to share the bundles of the last K blocks with different thread
+#[derive(Clone)]
+struct BundleStoredInLastK<Block: BlockT> {
+    #[allow(clippy::type_complexity)]
+    bundle_stored_in_last_k: Arc<Mutex<VecDeque<(Block::Hash, HashSet<Hash>)>>>,
+}
+
+impl<Block: BlockT> BundleStoredInLastK<Block> {
+    fn new() -> Self {
+        BundleStoredInLastK {
+            bundle_stored_in_last_k: Arc::new(Mutex::new(VecDeque::new())),
+        }
+    }
+
+    fn update(&self, incoming: VecDeque<(Block::Hash, HashSet<Hash>)>) {
+        *self.bundle_stored_in_last_k.lock() = incoming;
+    }
+}
+
+pub struct BundleValidator<Block: BlockT, Client> {
+    client: Arc<Client>,
+    bundle_syncer: BundleStoredInLastK<Block>,
+}
+
+impl<Block: BlockT, Client> Clone for BundleValidator<Block, Client> {
+    fn clone(&self) -> Self {
+        BundleValidator {
+            client: self.client.clone(),
+            bundle_syncer: self.bundle_syncer.clone(),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct SkipBundleValidation;
+
+pub trait ValidateBundle<Block: BlockT, DomainHash: Encode> {
+    // for consensus chain, check the duplicated bundle and receipts.
+    // for system domain, checks nothing.
+    fn validate_bundle(
+        &self,
+        at: BlockId<Block>,
+        signed_opaque_bundle: &SignedOpaqueBundle<NumberFor<Block>, Block::Hash, DomainHash>,
+    ) -> sp_blockchain::Result<()>;
+}
+
+impl<Block: BlockT, DomainHash: Encode> ValidateBundle<Block, DomainHash> for SkipBundleValidation {
+    fn validate_bundle(
+        &self,
+        _at: BlockId<Block>,
+        _signed_opaque_bundle: &SignedOpaqueBundle<NumberFor<Block>, Block::Hash, DomainHash>,
+    ) -> sp_blockchain::Result<()> {
+        Ok(())
+    }
+}
+
+impl<Block, DomainHash, Client> ValidateBundle<Block, DomainHash> for BundleValidator<Block, Client>
+where
+    Block: BlockT,
+    DomainHash: Encode,
+    Client: HeaderBackend<Block>,
+{
+    fn validate_bundle(
+        &self,
+        at: BlockId<Block>,
+        signed_opaque_bundle: &SignedOpaqueBundle<NumberFor<Block>, Block::Hash, DomainHash>,
+    ) -> sp_blockchain::Result<()> {
+        // The hash used here must be the same as what is maintaining in `bundle_stored_in_last_k`,
+        // namely the hash of `SignedOpaqueBundle`
+        let incoming_bundle = signed_opaque_bundle.hash();
+        let block_bundles = self.bundle_syncer.bundle_stored_in_last_k.lock();
```

Comment:
Since BundleStoredInLastK has an update() method, it would also be preferable to have a similar `is_bundle_present()` (or some such name). That way callers don't need to know about internal locking mechanism, and lock can be dropped sooner in the called function. Think @liuchengxu suggested something similar. 

If we go with the current way, please drop the lock sooner (using `std::mem::drop()`), after L238.

Reply:
Right, having a dedicated method for this check will be helpful.

Reply:
It doesn't seem that this lock needs to be held for this long, maybe wrap with `for` in `{}` to ensure it is dropped sooner?